README.md

```diff
@@ -25,3 +25,20 @@ The app deployed as part of this solution does the following:
 
 The agent crashes whenever it encounters an unrecoverable error with the metadata APIs.
 This agent is not production hardened yet and so use it with caution.
+
+## Graceful terminations for regular pods (Non-system pods)
+
+The pods that are not in the kube-system are called **regular pods** in this agent.
+By default, regular pods are deleted immediately before deleting system pods.
+If you want to delete regular pods gracefully, please add `--system-pod-grace-period=n` in arguments according to the following rules:
+
+- If targeted VM is Preemptible VM, specify `n` with a value from `0s` to `14s`.
+- If targeted VM is regular VM, specify `n` with a value from `0s` to the value of `(--regular-vm-timeout / 2) - 1`.
+
+If you follow the rules above, `VM timeout - system-grace-pod-period` will be given as a grace period for deleting regular pods.
+
+Also, `the timeout value of VM (e.g. preemptible=30s) / 2` cannot be used as a maximum value for the `--system-pod-grace-period`.
+This is because, the timing at which these delete events are triggered by a subscriber to GCE's Metadata API is different from the timing of the actual delete process.
+In other words, with proper consideration of the elapsed time, it's not possible to take advantage of all the duration remaining from the time a VM is known to be scheduled to delete until the VM is deleted actually.
```

Comment:
This is a little bit confusing. My understanding is the time between GCE metadata call happening and the call to delete pods should minimal since the GCE metadata call is using a [hanging get](https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/blob/master/termination/gceTerminationHandler.go#L157). 

My understanding is the main reason for this behavior is due to this [check](https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/blob/master/termination/eviction.go#L78-L80). The check of `2*p.systemPodGracePeriod` doesn't make a lot of sense to me... in my understanding we should be able to give a grace period to regular pods of (time between VM termination - systemPodGracePeriod) > 0.

Reply:
What I'm referring to here is after getting some event by the code you showed.
Let's organize a little.

First of all, after getting the event, `handleMaintenanceEvents` is triggered and the `storePendingTermination` is executed internally.
https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/blob/d56ae034967302518f8613d7333de5610965fd23/termination/gceTerminationHandler.go#L142

Furthermore, the `TerminationTime` is set by it.
https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/blob/d56ae034967302518f8613d7333de5610965fd23/termination/gceTerminationHandler.go#L116

It's expected that the processing up to this point will be executed synchronously.

Then, using the value set above, the notification for the termination will be passed through the channel managed in `Start`.
https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/blob/d56ae034967302518f8613d7333de5610965fd23/termination/handler.go#L96
From that, the timeout value used when actually evict in `processNodeState` is calculated.
At the very least, it's certain that the preemptible timeout at this point is less than 30 seconds.
https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/blob/d56ae034967302518f8613d7333de5610965fd23/termination/handler.go#L58

Reply:
Thanks for explanation. I think in the README, it might be ok to omit the last paragraph explanation since it doesn't provide too much additional clarity and is more an implementation detail of this library.

Reply:
With your advice, I considered removing it, but I suspect that users of this library might wonder why they can't specify a value of `30s/2` for regular pods. Considering the time mechanism described below, I think it might be more helpful to write the reason, but how about that?

Reply:
Thanks. My main point regarding the explanation is the reason that we don't allow 30s/2 isn't because

> "the timing at which these delete events are triggered by a subscriber to GCE's Metadata API is different from the timing of the actual delete process"

but rather because of the [check here](https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/blob/master/termination/eviction.go#L78-L80). If that check was removed for example, you could use almost all of the 30 seconds. 

This is kind of an implementation detail and longer term I think we should remove that check and reserve some amount of time to systempods (say 5-10 secs) and the rest of the time allocate to deletion of user's pods. That's why I think the explanation is maybe a little bit too much detail.

Reply:
That's fair, fully makes sense to me. Thanks. Let's keep the mention with dropping the reason: https://github.com/GoogleCloudPlatform/k8s-node-termination-handler/pull/29/commits/39c5088ae1b491e369ef8ce00be3f9c15479f0c9